packages/components/src/app/word-query/word-query-adapter.js

```diff
@@ -0,0 +1,115 @@
+
+    const cache = new InMemoryCache({
+      typePolicies: {
+        Query: {
+          fields: {
+            word: {
+              read: wordQueryResolver
+            }
+          }
+        }
+      }
+    })
+
+    this._gql = new ApolloClient({
+      cache,
+      link: new HttpLink({ uri: '/graphql', fetch }),
+      resolvers
+    })
+  }
+
+  static getInstance () {
+    if (!singleInstance) {
+      singleInstance = new WordQueryAdapter()
+    }
+    return singleInstance
+  }
+
+  /**
+   * An asynchronous GraphQL query. Data is returned when the query is fully completed.
+   *
+   * @param {object} options - Data that will be passed to the GraphQL query.
+   * @param {object} options.variables - Variables to be passed to the GraphQL query.
+   * @returns {Promise<*>} - A promise that is resolved with the GraphQL result.
+   */
+  async query ({ variables } = {}) {
+    const gqlQuery = {
+      query: WORD_QUERY,
+      variables
+    }
+    return this._gql.query(gqlQuery)
+  }
+
+  /**
+   * An asynchronous GraphQL query. Data is returned when the query is fully completed.
+   *
+   * @param {object} options - Data that will be passed to the GraphQL query.
+   * @param {object} options.variables - Variables to be passed to the GraphQL query.
+   * @param {Function} options.dataCallback - A function to be called every time new data arrives.
+   * @param {number} options.pollInterval - An interval in milliseconds specifying how often a new data will be retrieved.
+   * @param {number} options.maxIterations - How many iterations to do before stopping the query.
+   */
+  observableQuery ({ variables, dataCallback, pollInterval = 1000, maxIterations = 10 } = {}) {
+    let counter = 0
+    const observableQuery = this._gql.watchQuery({ query: WORD_QUERY, pollInterval, variables })
+    observableQuery.subscribe({
+      next: ({ data }) => {
+        const loading = data.word.state.loading
+        if (loading === false) {
+          observableQuery.stopPolling()
+        }
+        if (counter >= maxIterations) {
+          observableQuery.stopPolling()
+        }
+        counter++
+        dataCallback(data.word.homonyms, data.word.state, data.word.errors)
+      }
+    })
+  }
+
+  /**
+   * Similar to `observableQuery()`, but with its own callback implementation.
+   *
+   * @param {object} options - Data that will be passed to the GraphQL query.
+   * @param {object} options.variables - Variables to be passed to the GraphQL query.
+   * @param {number} options.pollInterval - An interval in milliseconds specifying how often a new data will be retrieved.
+   * @param {number} options.maxIterations - How many iterations to do before stopping the query.
+   * @returns {Promise<*>} - A promise that is resolved with GraphQL data or is rejected with an error.
+   */
+  async getFromObservableQuery ({ variables, pollInterval = 1000, maxIterations = 10 } = {}) {
+    return new Promise((resolve, reject) => {
```

Comment:
I think that it is pretty the same - define a method as async and return a new Promise, I believe that async is redundant.

And also I didn't finally understand the structure of this request:

we have `this.observableQuery`  that does the following steps:

(by the way I think that having the same names for a local variable and the whole method around of this variable is rather confusing and makes the code worse readable, also you don't even return the local variable as the result of this method)
 - define `observableQuery` as `watchQuery` of Appolo client that is equal to the `ObservableQuery`
 - define `subscribe` for this `observableQuery` and on received good results inits dataCallback with predefined data result (I believe you would add error handling here later)

and then here `getFromObservableQuery` - we redefine dataCallback with the different argumnets list?
Was this `data.word.homonyms, data.word.state, data.word.errors` and becomes this `(data, loading)` ?
Or may be I have mixed the same names for different intances?

Reply:
> I think that it is pretty the same - define a method as async and return a new Promise, I believe that async is redundant.

Yes, it is not required technically, but I've used it as a visual indication that this function executes asynchronously (i.e. returns a promise). Declaring the function `async` also changes the exception handling: any exception thrown and not caught is transformed into a rejection. However this feature is not used in the `getFromObservableQuery()`  as we have a `try/catch` block around all the code within this function.

It all is very opinionated. Here is a discussion about it on StackOverflow: https://stackoverflow.com/questions/44106907/when-to-mark-function-as-async I think we should make a decision about whether to use or not to use `async` in cases like that and stick to it.

@balmas, @irina060981, what do you think should be our approach here?

Reply:
> by the way I think that having the same names for a local variable and the whole method around of this variable is rather confusing and makes the code worse readable, also you don't even return the local variable as the result of this method

Agree with that, will rename the local variable.

Reply:
`observableQuery` is a wrapper around the Apollo Watcher that adds the following functionality:
* It introduces that maximum number of iterations that is used to terminate the long-running queries
* It checks our own custom `loading` status to verify if the query is still running
* It splits query response into three separate parts: homonyms, state, and errors thus hiding the internal structure of the GraphQL response from the client.

Reply:
> and then here getFromObservableQuery - we redefine dataCallback with the different argumnets list?
Was this data.word.homonyms, data.word.state, data.word.errors and becomes this (data, loading) ?

Oh, there is an error with the arguments inside the `getFromObservableQuery()`. I've missed it because this method is not used in the current implementation. I will update it to fix the issue. Thanks for catching this!

Reply:
> It all is very opinionated. Here is a discussion about it on StackOverflow: https://stackoverflow.com/questions/44106907/when-to-mark-function-as-async I think we should make a decision about whether to use or not to use `async` in cases like that and stick to it.
> 
> @balmas, @irina060981, what do you think should be our approach here?

I'm having deja vu about this one. I think we've debated about it before. I prefer to see the async keyword.

Reply:
> I'm having deja vu about this one. I think we've debated about it before.

I think we can start a "Best practices" document to put cases like that there so we won't forget. This could be the first case for it 🙂.

Reply:
I agree, that it could be used with directly defined Promise in return and in the case we return non-Promise, that would be converted to Promise by async keyword.
Thank you for the explanations.

Reply:
I will start to write the Best Practices document.